File: 647-palindromic-substrings/my_approach.cpp
// Time complexity: O(n^2), Space complexity: O(n^2)
class Solution {
public:
    int countSubstrings(string s) {
        int len = s.length();
        vector<vector<int>> dp(len, vector<int>(len, 0));
        int numSubs = 0;
        
        for (int i = 0; i < len; i++) {
            dp[i][i] = 1;
        }

        for (int i = len - 1; i >= 0; i--) {
            for (int j = i; j < len; j++) {
                if (i == j) {
                    numSubs += 1;
                }
                else if (i == j  - 1){
                    if (s[i] == s[j]) {
                        dp[i][j] = 2;
                        numSubs += 1;
                    }
                }
                else {
                    if (s[i] == s[j] && dp[i + 1][j - 1] == j - i - 1) {
                        dp[i][j] = j - i + 1;
                        numSubs += 1;
                    }
                }
            }
        }

        return numSubs;
    }
};

File: 0003-longest-substring-without-repeating-characters/my_approach.py
# Time complexity: O(N), Space complexity: O(N)
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        chMap = {}
        size = len(s)
        l = r = 0
        max_len = -1e9

        while r < size:
            ch = s[r]
            if ch in chMap:
                l = max(l, chMap[ch] + 1)
            chMap[ch] = r

            max_len = max(max_len, r - l + 1)
            r+= 1
        
        if max_len == -1e9:
            return size

        return max_len